lightningd/invoice.c

```diff
@@ -356,33 +388,51 @@ static struct route_info **select_inchan(const tal_t *ctx,
 			continue;
 		}
 
-		/* Even after reserve taken into account */
-		if (!amount_msat_sub_sat(&avail,
-					 avail, c->our_config.channel_reserve))
+		/* Even after taken into account their reserve */
+		if (!amount_msat_sub_sat(&capacity_to_pay_us, their_msat,
+				c->our_config.channel_reserve))
 			continue;
 
-		if (!amount_msat_sub(&excess, avail, capacity_needed))
+		if (!amount_msat_sub(&excess, capacity_to_pay_us, amount_needed))
 			continue;
 
+		/* Excess as percentage of capacity, assumes check_config_bounds */
+		excess_perc = (double)excess.millisatoshis * 0.1 /               /* Raw:excess_perc */
+			  (c->funding.satoshis                                       /* Raw:excess_perc */
+			  - c->our_config.channel_reserve.satoshis                   /* Raw:excess_perc */
+			  - c->channel_info.their_config.channel_reserve.satoshis);  /* Raw:excess_perc */
+
```

Comment:
OK, with that many /* Raw */ overrides is a big red flag.  We do overflow checking to ensure robustness, and we should keep that:

How about:
```
   struct amount_msat total_capacity;

   if (!amount_sat_to_msat(&total_capacity, c->funding)
       || !amount_msat_sub_sat(&total_capacity, c->our_config.channel_reserve)
       || !amount_msat_sub_sat(&total_capacity, c->channel_info.their_config.channel_reserve)) {
           log_broken(lg->log, "Channel %s capacity %s overflow (reserves %s + %s)!",
                          type_to_string(tmpctx, struct short_channel_id, &c->scid),
                          type_to_string(tmpctx, struct amount_sat, &c->funding),
...;
           continue;
    }

    excess_perc = (double)excess.millisatoshis / total_capacity.millisatoshis; /* Raw: double fraction */
```

I suggest changing excess_perc to excess_fraction; the 100 multiplier is just a distraction AFAICT?

Reply:
Agreed, I see what you mean. Then a small point: in the code base I couldn't find consistent use of the term `capacity`, [sometimes](https://github.com/ElementsProject/lightning/blob/77236caa91bc548d24d5e886bcaf659cc1fd4c77/gossipd/routing.c#L65) it's synonym for `funding`, and other [times](https://github.com/ElementsProject/lightning/blob/77236caa91bc548d24d5e886bcaf659cc1fd4c77/openingd/openingd.c#L201) for `funding` minus `reserves`.

The later seems more appropriate to me, which would then make `total_capacity` synonym to `funding` (or just call it `funding`).

Reply:
Hope that is ok.